### PR TITLE
Ensure support of the transport-nio by security plugin (HTTP)

### PR DESCRIPTION
**Pull request by reta**
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Ensure support of the transport-nio by security plugin (HTTP only), changes in `opensearch.yml`:
```
http.type: nio-http-transport-secure
```

```
$ curl -kv https://localhost:9200/ -u admin:_ad0m#Ns_ --http1.1                                                                                                                                                                                                                              08:32:15 [39/437]
* Host localhost:9200 was resolved.                                                  
* IPv6: ::1                                                                          
* IPv4: 127.0.0.1                                                                    
*   Trying [::1]:9200...                                                             
* Connected to localhost (::1) port 9200                                             
* ALPN: curl offers http/1.1                                                         
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256 / x25519 / RSASSA-PSS
* ALPN: server accepted http/1.1                                                     
* using HTTP/1.x                                                                                                                                                                                                                                                                                                                                      
* Server auth using Basic with user 'admin'                                                                                                                                                                                                                                                                                                           
> GET / HTTP/1.1                                                                                                                                                                                                                                                                                                                                      
> Host: localhost:9200                                                               
> Authorization: Basic YWRtaW46X2FkMG0jTnNf
> User-Agent: curl/8.9.1                                                             
> Accept: */*                                                                        
>                                                                                    
* Request completely sent off                                                        
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
< HTTP/1.1 200 OK                                                                    
< X-OpenSearch-Version: OpenSearch/3.0.0 (opensearch)
< content-type: application/json; charset=UTF-8
< content-length: 576                                                                
<                                                                                    
{                                                                                    
  "name" : "host",                                              
  "cluster_name" : "opensearch",                                                     
  "cluster_uuid" : "ByTSGP1rSGOnak0LKonwWw",
  "version" : {                                                                      
    "distribution" : "opensearch",                                                   
    "number" : "3.0.0",                                                              
    "build_type" : "tar",                                                            
    "build_hash" : "9dd1a59847ad8c2716d002716521ac40afc69355",
    "build_date" : "2024-10-24T04:39:37.808432550Z",
    "build_snapshot" : false,                                                        
    "lucene_version" : "9.12.0",                                                     
    "minimum_wire_compatibility_version" : "2.18.0",
    "minimum_index_compatibility_version" : "2.0.0"
  },                                                                                 
  "tagline" : "The OpenSearch Project: https://opensearch.org/"
} 
```

### Related Issues
Closes https://github.com/opensearch-project/OpenSearch/issues/13245
<!-- List any other related issues here -->

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugins/transport-reactor-netty4/src/test/java/org/opensearch/http/reactor/netty4/ReactorHttpClient.java [hunk 1]
```diff
@@ -181,7 +181,7 @@ private List<FullHttpResponse> processRequestsWithBody(
     private List<FullHttpResponse> sendRequests(
         final InetSocketAddress remoteAddress,
         final Collection<FullHttpRequest> requests,
-        boolean orderer
+        boolean ordered
```
Comment (reta) [inline plugins/transport-reactor-netty4/src/test/java/org/opensearch/http/reactor/netty4/ReactorHttpClient.java:184]: Fixing minor typo

**Comment by reta** [comment]
@cwperks  @andrross @dblock would really appreciate if you folks could take a look, thanks!

File: plugins/transport-nio/src/internalClusterTest/java/org/opensearch/http/nio/NioPipeliningIT.java [hunk 2]
```diff
@@ -61,7 +61,7 @@ public void testThatNioHttpServerSupportsPipelining() throws Exception {
         TransportAddress[] boundAddresses = httpServerTransport.boundAddress().boundAddresses();
         TransportAddress transportAddress = randomFrom(boundAddresses);
 
-        try (NioHttpClient nettyHttpClient = new NioHttpClient()) {
+        try (NioHttpClient nettyHttpClient = NioHttpClient.http()) {
```
Comment (cwperks) [inline plugins/transport-nio/src/internalClusterTest/java/org/opensearch/http/nio/NioPipeliningIT.java:64]: nit: Should this be renamed to `nioHttpClient`?

**Review by cwperks — COMMENTED** [review]
Nice work @reta! I'll take a few passes at this PR and get feedback ASAP. 

There was a pretty sizable change to the netty pipeline in 2.11 in these 2 PRs which should be extended to any http transport.

- https://github.com/opensearch-project/OpenSearch/pull/10261
- https://github.com/opensearch-project/security/pull/3418

There is an automated test [here](https://github.com/opensearch-project/security/blob/main/src/integrationTest/java/org/opensearch/security/ResourceFocusedTests.java) which performs a stress test.

**Comment by reta** [comment]
> Nice work @reta! I'll take a few passes at this PR and get feedback ASAP.

Thanks @cwperks !

> There was a pretty sizable change to the netty pipeline in 2.11 in these 2 PRs which should be extended to any http transport.

I think these improvements could go independently of each other, right?

**Comment by cwperks** [comment]
I think we need to get that test passing w/ the security plugin to declare that this transport works w/ the security plugin. You can run it with the changes in this branch: https://github.com/cwperks/security/pull/33

**Comment by cwperks** [comment]
Thank you @reta! This PR looks good to me. I'm not a maintainer of this repo so I will leave a comment approval :shipit:

**Review by andrross — APPROVED** [review]

**Comment by andrross** [comment]
Just curious @reta, is there any reason to prefer `transport-nio` over the Netty version? I'm wondering if this is something to consider removing in the next major version.

**Comment by reta** [comment]
> Just curious @reta, is there any reason to prefer `transport-nio` over the Netty version? I'm wondering if this is something to consider removing in the next major version.

Thanks @andrross  I think we've never benchmarked / compared these transports side by side. I will try to research the history of nio transport, thanks!

**Comment by andrross** [comment]
Thanks @reta. I'm always looking for opportunity to remove code!

**Comment by reta** [comment]
> Thanks @reta. I'm always looking for opportunity to remove code!

@andrross traced it back to https://github.com/elastic/elasticsearch/issues/27260 (which basically what I also suspected):

> We are interested in introducing alternative transport options that do not depend on external libraries. From time to time we have had some friction with using netty related to buffer pooling, oom excetions, exception handling, unsafe, etc. We want to look explore if a homegrown transport could potentially meet our needs.

Not sure it is relevant these day, unless `java.nio` shines in some flows

**Pull request merged**